apache2/msc_reqbody.c

```diff
@@ -461,37 +461,30 @@ apr_status_t modsecurity_request_body_to_stream(modsec_rec *msr, const char *buf
         if(data == NULL)
             return -1;
 
-        memset(data, 0, msr->stream_input_length + 1 - buflen);
         memcpy(data, msr->stream_input_data, msr->stream_input_length - buflen);
+        data[msr->stream_input_length - buflen] = '\0';
 
         stream_input_body = (char *)realloc(msr->stream_input_data, msr->stream_input_length + 1);
 
         msr->stream_input_data = (char *)stream_input_body;
     }
 
     if (msr->stream_input_data == NULL) {
-        if(data)    {
-            free(data);
-            data = NULL;
-        }
+        if (data) free(data);
         *error_msg = apr_psprintf(msr->mp, "Unable to allocate memory to hold request body on stream. Asked for %" APR_SIZE_T_FMT " bytes.",
                 msr->stream_input_length + 1);
         return -1;
     }
 
-    memset(msr->stream_input_data, 0, msr->stream_input_length+1);
-
     if(first_pkt)   {
         memcpy(msr->stream_input_data, buffer, msr->stream_input_length);
     } else {
         memcpy(msr->stream_input_data, data, msr->stream_input_length - buflen);
         memcpy(msr->stream_input_data+(msr->stream_input_length - buflen), buffer, buflen);
     }
+    msr->stream_input_data[msr->stream_input_length] = '\0';
```

Comment:
If I'm right the target string's length can be 2 different value. But here it handled only the first case - will this be good?

Reply:
First case, we copy `msr->stream_input_length` bytes.
Second case, we copy `msr->stream_input_length - buflen` bytes, then `buflen` bytes.
So, `msr->stream_input_length` bytes in both cases.

Reply:
Okay, now I reviewed the whole logic and I see. Thanks.